extensions/notification/NotificationQueuePopulator.js

```diff
@@ -119,22 +119,75 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
         return this.bnConfigManager.removeConfig(bucketName);
     }
 
+    /**
+     * Returns the correct versionId
+     * to display according to the
+     * versioning state of the object
+     * @param {Object} value log entry object
+     * @param {Object} overheadFields - extra fields missing from the log entry
+     * @return {String} versionId
+     */
+    _getVersionId(value, overheadFields) {
+        const versionId = value.versionId || (overheadFields && overheadFields.versionId);
+        const isNullVersion = value.isNull;
+        const isVersioned = !!versionId;
+        // Versioning suspended objects have
+        // a versionId, however it is internal
+        // and should not be used to get the object
+        if (isNullVersion || !isVersioned) {
+            return null;
+        }
+        return versionId;
+    }
+
+    /**
+     * Decides if we should process the entry.
+     * Since we get both master and version events,
+     * we need to avoid pushing two notifications for
+     * the same event.
+     * - For non-versioned buckets, we process the master
+     * objects' events.
+     * - For versioned buckets, we process version events
+     * and ignore all master events.
+     * - For versioning suspended buckets, we need to process
+     * both master and version events, as the master is considered
+     * a separate version.
+     * @param {String} key object key
+     * @param {Object} value object metadata
+     * @return {boolean} - true if entry is valid
+     */
+    _shouldProcessEntry(key, value) {
+        const isMaster = isMasterKey(key);
+        const hasVersionId = !!value.versionId;
+        const isVersionedMaster = isMaster && hasVersionId;
+        const isNullVersion = isMaster && value.isNull;
+        const isNullWithVersionId = hasVersionId && value.isNull;
```

Comment:
To exclude the null keys from this, we also need to check the key itself: if it ends with `\0`, it is a null key and should not be excluded (because then we want to process that entry as there is no master key associated). Null keys have a `versionId` field in their metadata because they correspond to null versions from a versioning-suspended bucket, or in V1 they may also be a current delete marker on a versioned bucket.